### PR TITLE
feat(eslint-plugin-fiori-tools): add V4 table personalization rule

### PR DESCRIPTION
**Pull request by AlinaGovoruhina**
Add an eslint rule: check table personalization is enabled in OData V4 projects.
Checks application manifest.json file for all available tables.

Valid values: undefined, true, all available personalization properties set to true.

Invalid: false, properties set to false or undefined (In case of using an object, omitted properties are considered disabled).

Quick fix: sets personalization to true (enables all table personalization features).
Note: table grouping is available for Analytical and Responsive tables only.

### CODE REVIEW

File: packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts [hunk 1]
```diff
@@ -0,0 +1,862 @@
...
+                            'targets',
+                            'IncidentsObjectPage',
+                            'options',
+                            'settings',
+                            'controlConfiguration',
+                            'incidentFlow/@com.sap.vocabularies.UI.v1.LineItem',
+                            'tableSettings',
+                            'personalization'
+                        ],
+                        value: false
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts]: When the test description is correct, value should be {}

File: packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts [hunk 2]
```diff
@@ -202,6 +202,41 @@ describe('createJsonFixer', () => {
             expect(fixer.replaceTextRange).toHaveBeenCalled();
             expect(result).toMatchSnapshot();
         });
+
+        it('should update the value for a parent node - fixParent = true', () => {
+            const sourceText = `{"parent": {"old": true}}`;
+            const { node, ast } = createNodeFromJson(sourceText, ['parent']);
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts]: Shouldn't we use ['parent', 'old'] for the node path here so that the node path matches the validationPath below?

Same question for the two following tests...

File: packages/eslint-plugin-fiori-tools/docs/rules/sap-table-personalization.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+}
+```
+
+Every table `personalization` property is correctly set to `true`.
+Omitting table personalization from the `manifest.json` table settings is also correct, because all personalization settings are provided by default for all tables.
+
+## How to Fix
+
+To fix the warning, ensure that either the `personalization` property is set to `true` or not defined in the table settings.
+
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/docs/rules/sap-table-personalization.md:101]: Do we also want to mention here what the automatic fix does?

File: packages/eslint-plugin-fiori-tools/docs/rules/sap-table-personalization.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+```json
+{
+  "tableSettings": {
+    "personalization": {}
+  }
+}
+```
+
+If the value "object" is used, omitting a setting is treated as `false`.
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/docs/rules/sap-table-personalization.md]: Shall we also have a partially defined object as an example?

And also an example with individual properties set to false?

File: packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts [hunk 5]
```diff
@@ -0,0 +1,862 @@
...
+            [FACETSV4]
+        ),
+        createValidTest(
+            {
+                name: 'V4 - list report page - ResponsiveTable + minUI5=1.120 - group is true',
+                filename: V4_MANIFEST_PATH,
+                code: getManifestAsCode(V4_MANIFEST, [
+                    {
+                        path: ['sap.ui5', 'dependencies', 'minUI5Version'],
+                        value: '1.120.1'
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts]: Do we also need a test for exact boundaries like '1.120.0'? Maybe we can use that version here.
Same for '1.108.0' below.

File: packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts [hunk 6]
```diff
@@ -0,0 +1,862 @@
...
+                            'tableSettings',
+                            'personalization'
+                        ],
+                        value: true
+                    }
+                ])
+            },
+            [FACETSV4]
+        ),
+        createInvalidTest(
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts:920]: Test coverage is quiet good.
Do we also need a test for a single undefined property to get the 'is disabled' message?

File: packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts (hunk 2 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts]: Do you mean in `createNodeFromJson()`? This creates a node from 'parent' with 'old' inside, if I pass it ['parent', 'old'] it will create a node from 'old'.

File: packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts (hunk 1 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts]: Setting `personalization` to false is the same as setting every property to false, so everything is disabled. `{}` also sets all properties to `false`, there is a test for empty object value below.

File: packages/eslint-plugin-fiori-tools/docs/rules/sap-table-personalization.md (hunk 3 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/docs/rules/sap-table-personalization.md:101]: Added fix information here.

File: packages/eslint-plugin-fiori-tools/docs/rules/sap-table-personalization.md (hunk 4 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/docs/rules/sap-table-personalization.md]: Added this code example.

File: packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts (hunk 5 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts]: Good point, added tests for AnalyticalTable+1.108.0 and ResponsiveTable 1.120.0, where `group` property value is available and checked.

File: packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts (hunk 6 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts:920]: Added the unit test for single undefined property.

File: packages/eslint-plugin-fiori-tools/src/rules/sap-table-personalization.ts [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+        problems.push({
+            type: TABLE_PERSONALIZATION,
+            pageName: page.targetName,
+            messageId: MessageIdByProperty[''],
+            manifest: {
+                uri: parsedApp.manifest.manifestUri,
+                object: parsedApp.manifestObject,
+                propertyPath: table.configuration.personalization.configurationPath
+            }
+        });
```
Comment (marufrasully) [inline packages/eslint-plugin-fiori-tools/src/rules/sap-table-personalization.ts:202]: returning here would save else block :)

**Review by marufrasully — APPROVED** [review]
Thanks @AlinaGovoruhina 
- code changes looks good
- great test coverage
- did not verify it

File: packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts (hunk 1 above)
Comment (marcovieth) [reply packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts]: But then the test description does not match the value "false", so it should mention "false".

File: packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts (hunk 2 above)
Comment (marcovieth) [reply packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts]: Yes, I mean `createNodeFromJson(sourceText, ['parent', 'old'])`.
My understanding is that it will then create a node from "old". Then with "fixParent: true" it will operate on the parent object (even if it's just a simulation). And node path = validation path, as it is with other tests.

File: packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts (hunk 1 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/test/rules/sap-table-personalization.test.ts]: Ok, I changed the tested value to be consistent with test name.

File: packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts (hunk 2 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts]: I see, I think I mixed up node and ast, so it is correct: we need `old` node fo the test. it is updated.

File: packages/eslint-plugin-fiori-tools/src/rules/sap-table-personalization.ts (hunk 7 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/src/rules/sap-table-personalization.ts:202]: Thanks, I added the return statement in https://github.com/SAP/open-ux-tools/pull/4104/commits/3e0ce1f1948fdd83123dd068a052737d485dd586

File: packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts [hunk 8]
```diff
@@ -489,6 +568,45 @@ describe('createJsonFixer', () => {
 
             expect(result).toEqual([]);
         });
+
+        it('should delete parent object value if fixParent = true', () => {
+            const sourceText = `{"parent": {"old": true}}`;
+            const { node, ast } = createNodeFromJson(sourceText, ['parent']);
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts]: Test case for "update" is ok now. I think for "delete" we also need: createNodeFromJson(sourceText, ['parent', 'old']);

File: packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts [hunk 9]
```diff
@@ -323,6 +362,46 @@ describe('createJsonFixer', () => {
             expect(result).toEqual([]);
             expect(fixer.insertTextBeforeRange).not.toHaveBeenCalled();
         });
+
+        it('should insert value on a parent node - fixParent = true', () => {
+            const sourceText = `{"parent": {"old": true}}`;
+            const { node, ast } = createNodeFromJson(sourceText, ['parent']);
```
Comment (marcovieth) [inline packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts]: Test case for "update" is ok now. I think for "insert" we also need: createNodeFromJson(sourceText, ['parent', 'old']);

File: packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts (hunk 8 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts]: Thanks, updated this test

File: packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts (hunk 9 above)
Comment (AlinaGovoruhina) [reply packages/eslint-plugin-fiori-tools/test/language/rule-fixer.test.ts]: Updated the test as well
